File: ios/BockMedia/Domain/SearchQueryCache.swift
import Foundation

/// Short-TTL in-memory cache so backspace/retype avoids repeat network calls.
enum SearchQueryCache {
    private static let ttlMs: Int64 = 45_000
    private static var query = ""
    private static var response: SearchResponse?
    private static var savedAtMs: Int64 = 0

    static func get(_ q: String) -> SearchResponse? {
        let trimmed = q.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return nil }
        guard query == trimmed, let response else { return nil }
        guard Int64(Date().timeIntervalSince1970 * 1000) - savedAtMs <= ttlMs else { return nil }
        return response
    }

    static func prefixExtension(_ q: String) -> SearchResponse? {
        let trimmed = q.trimmingCharacters(in: .whitespaces)
        guard trimmed.count >= 2 else { return nil }
        guard let hit = response else { return nil }
        let cached = query
        guard cached.count >= 1, trimmed != cached else { return nil }
        guard Int64(Date().timeIntervalSince1970 * 1000) - savedAtMs <= ttlMs else { return nil }
        guard trimmed.hasPrefix(cached) else { return nil }
        return hit
    }

    static func put(_ q: String, _ response: SearchResponse) {
        let trimmed = q.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return }
        query = trimmed
        self.response = response
        savedAtMs = Int64(Date().timeIntervalSince1970 * 1000)
    }

    static func invalidate() {
        query = ""
        response = nil
        savedAtMs = 0
    }
}

File: android/app/src/main/kotlin/com/bockmedia/console/domain/model/SearchQueryCache.kt
package com.bockmedia.console.domain.model

import com.bockmedia.console.data.api.dto.SearchResponse

/** Short-TTL in-memory cache so backspace/retype and tab switches avoid repeat network calls. */
object SearchQueryCache {
    private const val TTL_MS = 45_000L

    @Volatile private var query: String = ""
    @Volatile private var response: SearchResponse? = null
    @Volatile private var savedAtMs: Long = 0L

    fun get(q: String): SearchResponse? {
        val trimmed = q.trim()
        if (trimmed.isEmpty()) return null
        val hit = response ?: return null
        if (this.query != trimmed) return null
        if (System.currentTimeMillis() - savedAtMs > TTL_MS) return null
        return hit
    }

    /** Reuse a shorter cached query while the user extends the string (e.g. learn → learn t). */
    fun getPrefixExtension(q: String): SearchResponse? {
        val trimmed = q.trim()
        if (trimmed.length < 2) return null
        val hit = response ?: return null
        val cached = this.query
        if (cached.length < 2 || trimmed == cached) return null
        if (System.currentTimeMillis() - savedAtMs > TTL_MS) return null
        if (!trimmed.startsWith(cached)) return null
        return hit
    }

    fun put(q: String, response: SearchResponse) {
        val trimmed = q.trim()
        if (trimmed.isEmpty()) return
        this.query = trimmed
        this.response = response
        savedAtMs = System.currentTimeMillis()
    }

    fun invalidate() {
        query = ""
        response = null
        savedAtMs = 0L
    }
}
